File: tuan2/PhanC/bai47.py
import sys
import math

def khoang_cach(u, v):
    tong = sum((u[k] - v[k]) ** 2 for k in range(len(u)))
    return math.sqrt(tong)

def main():
    try:
        with open("hoa30.txt", "r") as f:
            lines = [line.strip() for line in f if line.strip()]
    except FileNotFoundError:
        print("Khong the mo tep hoa30.txt", file=sys.stderr)
        return

    first_line = lines[0].split()
    n = int(first_line[0])
    d = int(first_line[1])

    D = []
    for i in range(1, n + 1):
        parts = lines[i].split()
        features = [float(x) for x in parts[:d]]
        nhan = parts[d]
        D.append((i, features, nhan))

    input_data = sys.stdin.read().split()
    if not input_data:
        return

    idx = 0
    while idx < len(input_data):
        u = [float(x) for x in input_data[idx:idx + d]]
        idx += d

        kc_min = float("inf")
        du_doan = None
        id_gan_nhat = -1

        for mid, x, nhan in D:
            dist = khoang_cach(u, x)
            if dist < kc_min:
                kc_min = dist
                du_doan = nhan
                id_gan_nhat = mid

        print(f"{du_doan} — mau {id_gan_nhat}, khoang cach {kc_min:.4f}")

if __name__ == "__main__":
    main()

File: tuan2/PhanC/bai47.jl
using Printf

struct Mau
    id::Int
    x::Vector{Float64}
    nhan::String
end

function khoang_cach(u::Vector{Float64}, v::Vector{Float64})
    return sqrt(sum((u .- v) .^ 2))
end

function main()
    if !isfile("hoa30.txt")
        println(stderr, "Khong the mo tep hoa30.txt")
        return
    end

    lines = filter(!isempty, strip.(readlines("hoa30.txt")))
    first_line = split(lines[1])
    n = parse(Int, first_line[1])
    d = parse(Int, first_line[2])

    D = Vector{Mau}(undef, n)
    for i in 1:n
        parts = split(lines[i + 1])
        x = [parse(Float64, parts[j]) for j in 1:d]
        nhan = String(parts[d + 1])
        D[i] = Mau(i, x, nhan)
    end

    input_tokens = split(join(readlines(), " "))
    if isempty(input_tokens)
        return
    end

    idx = 1
    while idx <= length(input_tokens)
        u = [parse(Float64, input_tokens[idx + j - 1]) for j in 1:d]
        idx += d

        kc_min = Inf
        du_doan = ""
        id_gan_nhat = -1

        for m in D
            dist = khoang_cach(u, m.x)
            if dist < kc_min
                kc_min = dist
                du_doan = m.nhan
                id_gan_nhat = m.id
            end
        end

        @printf("%s — mau %d, khoang cach %.4f\n", du_doan, id_gan_nhat, kc_min)
    end
end

main()